### PR TITLE
Update some modules, Deprecated code, and fix testing for PHP 8.2

### PR DESCRIPTION
**Pull request by fiko**
### Description (*)
- Update some modules for PHP 8.2
    - laminas/laminas-captcha [2.16.0](https://github.com/laminas/laminas-captcha/releases/tag/2.16.0)
    - laminas/laminas-code [4.10.0](https://github.com/laminas/laminas-code/releases/tag/4.10.0)
    - laminas/laminas-db [2.17.0](https://github.com/laminas/laminas-db/releases/tag/2.17.0)
    - laminas/laminas-di [3.12.0](https://github.com/laminas/laminas-di/releases/tag/3.12.0)
    - laminas/laminas-router [3.11.1](https://github.com/laminas/laminas-router/releases/tag/3.11.1)
    - laminas/laminas-view [2.27.0](https://github.com/laminas/laminas-view/releases/tag/2.27.0)
    - wikimedia/less.php [v3.2.1](https://github.com/wikimedia/less.php/releases/tag/v3.2.1)
    - laminas/laminas-diactoros [2.24.0](https://github.com/laminas/laminas-diactoros/releases/tag/2.24.0)
    - phpspec/prophecy [1.17.0](https://github.com/phpspec/prophecy/releases/tag/v1.17.0)
- Update some deprecated code for PHP 8.2.
- Fix unit testing result.

### Manual testing scenarios (*)
1. clone magento,
2. composer install using PHP 8.2
3. bin/magento setup:upgrade && bin/magento setup:di:compile && bin/magento setup:static-content:deploy

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by fiko** [comment]
@magento run all tests

File: composer.json [hunk 1]
```diff
@@ -9,7 +9,8 @@
     "config": {
         "allow-plugins": {
             "dealerdirect/phpcodesniffer-composer-installer": true,
-            "magento/*": true
+            "magento/*": true,
+            "php-http/discovery": true
```
Comment (ihor-sviziev) [inline composer.json]: Could you please explain why it was added?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
In general, your changes look good to me.
Could you please fix the failing static tests?

File: composer.json [hunk 2]
```diff
@@ -41,39 +42,41 @@
...
         "laminas/laminas-modulemanager": "^2.11",
         "laminas/laminas-mvc": "^3.3",
-        "laminas/laminas-oauth": "^2.4",
+        "laminas/laminas-oauth": "^2.5",
         "laminas/laminas-permissions-acl": "^2.10",
+        "laminas/laminas-recaptcha": "^3.6",
         "laminas/laminas-servicemanager": "^3.16",
-        "laminas/laminas-soap": "^2.10",
+        "laminas/laminas-session": "^2.16",
+        "laminas/laminas-soap": "^2.11",
```
Comment (fredden) [inline composer.json]: What's not compatible with version `2.10` in this pull request?
```suggestion
        "laminas/laminas-soap": "^2.10",
```

File: composer.json [hunk 3]
```diff
@@ -41,39 +42,41 @@
...
+        "laminas/laminas-file": "^2.12",
         "laminas/laminas-filter": "^2.17",
         "laminas/laminas-http": "^2.15",
         "laminas/laminas-i18n": "^2.17",
         "laminas/laminas-mail": "^2.16",
         "laminas/laminas-mime": "^2.9",
         "laminas/laminas-modulemanager": "^2.11",
         "laminas/laminas-mvc": "^3.3",
-        "laminas/laminas-oauth": "^2.4",
+        "laminas/laminas-oauth": "^2.5",
```
Comment (fredden) [inline composer.json]: What's not compatible with version `2.4` in this pull request?
```suggestion
        "laminas/laminas-oauth": "^2.4",
```

File: composer.json [hunk 4]
```diff
@@ -41,39 +42,41 @@
         "elasticsearch/elasticsearch": "^7.17||^8.5",
         "ezyang/htmlpurifier": "^4.14",
         "guzzlehttp/guzzle": "^7.4",
-        "laminas/laminas-captcha": "^2.12",
-        "laminas/laminas-code": "^4.5",
-        "laminas/laminas-db": "^2.15",
-        "laminas/laminas-di": "^3.7",
+        "laminas/laminas-captcha": "^2.16",
```
Comment (fredden) [inline composer.json]: What's not compatible with version `2.12` in this pull request?
```suggestion
        "laminas/laminas-captcha": "^2.12",
```

File: composer.json [hunk 5]
```diff
@@ -41,39 +42,41 @@
...
+        "laminas/laminas-code": "^4.8",
```
Comment (fredden) [inline composer.json]: What's not compatible with version `4.5` in this pull request?
```suggestion
        "laminas/laminas-code": "^4.5",
```

File: composer.json [hunk 6]
```diff
@@ -41,39 +42,41 @@
...
+        "laminas/laminas-db": "^2.17",
```
Comment (fredden) [inline composer.json]: What's not compatible with version `2.15` in this pull request?
```suggestion
        "laminas/laminas-db": "^2.15",
```

File: composer.json [hunk 7]
```diff
@@ -41,39 +42,41 @@
...
+        "laminas/laminas-di": "^3.12",
```
Comment (fredden) [inline composer.json]: What's not compatible with version `3.7` in this pull request?
```suggestion
        "laminas/laminas-di": "^3.7",
```

File: composer.json [hunk 8]
```diff
@@ -41,39 +42,41 @@
...
         "laminas/laminas-escaper": "^2.10",
         "laminas/laminas-eventmanager": "^3.5",
         "laminas/laminas-feed": "^2.17",
-        "laminas/laminas-file": "^2.11",
+        "laminas/laminas-file": "^2.12",
```
Comment (fredden) [inline composer.json]: What's not compatible with version `2.11` in this pull request?
```suggestion
        "laminas/laminas-file": "^2.11",
```

File: composer.json [hunk 9]
```diff
@@ -41,39 +42,41 @@
...
+        "magento/composer": "^1.9",
         "magento/composer-dependency-version-audit-plugin": "^0.1",
         "magento/magento-composer-installer": ">=0.4.0-beta1",
         "magento/zend-cache": "^1.16",
         "magento/zend-db": "^1.16",
         "magento/zend-pdf": "^1.16",
         "monolog/monolog": "^2.7",
         "opensearch-project/opensearch-php": "^1.0 || ^2.0",
-        "pelago/emogrifier": "^6.0.0",
+        "pelago/emogrifier": "^7.0",
```
Comment (fredden) [inline composer.json]: What's not compatible with version `6.0` in this pull request?
```suggestion
        "pelago/emogrifier": "^6.0 || ^7.0",
```

File: composer.json [hunk 10]
```diff
@@ -86,7 +89,7 @@
         "tubalmartin/cssmin": "^4.1",
         "web-token/jwt-framework": "^3.1",
         "webonyx/graphql-php": "^14.11",
-        "wikimedia/less.php": "^3.0"
+        "wikimedia/less.php": "^4.0"
```
Comment (fredden) [inline composer.json]: What's not compatible with version `3.0` in this pull request?
```suggestion
        "wikimedia/less.php": "^3.0 || ^4.0"
```

**Comment by hostep** [comment]
The 2.4.6 code [was merged 2 hours ago](https://github.com/magento/magento2/commit/176e388fb6168f76d1d8531b2375d72aaa00ba81) in the `2.4-develop` branch which already contains a bunch of fixes for PHP 8.2. So you'll have to resolve merge conflicts first to see if some of these fixes weren't already included in 2.4.6 and are still needed.

**Comment by fiko** [comment]
thanks for the review, will try to fix the conflict and testing

**Comment by fiko** [comment]
@magento run all tests

File: composer.json (hunk 1 above)
Comment (fiko) [reply composer.json]: hi @ihor-sviziev 

it will auto install PSR, here's for the documentation of this https://packagist.org/packages/php-http/discovery

**Comment by fiko** [comment]
I tested Static Tests locally and faced the error that I need to use onlyMethods instead of addMethods, but when I test it on github, it requires addMethods.

will try to fix it and rely on github static tests instead of locally.

File: composer.json (hunk 10 above)
Comment (mrtuvn) [reply composer.json:94]: seem up to v4 bring more fixes for packages. I thinks we safe to use v4 with mininum support php v 7.4+

Comment (mrtuvn) [reply composer.json]: https://gerrit.wikimedia.org/g/mediawiki/libs/less.php/+log/v4.0.0

Comment (fiko) [reply composer.json]: Hey,

it was upgraded when magento 2.4.6 was not released and when I `setup:static-content:deploy`, it returns error (apology I hadn't take a screenshot of it).

once I get latest magento 2.4.6, it is working, so will revert it back to ^3.0

thanks for the review

**Comment by fiko** [comment]
@magento run Static Tests

File: composer.json (hunk 10 above)
Comment (mrtuvn) [reply composer.json]: "wikimedia/less.php": "^3.2"
current we using this ver. If you are planning to upgrade to v4 you should consider re-check less compilation by run setup static content deploy to make sure everything still works fine

Comment (fiko) [reply composer.json]: I already tried, I tried deploy using version ^3.2 and version ^4.0, everything  was fine. is it fine to upgrade `wikimedia/less.php:^4.0` on `lib/internal/Magento/Framework/composer.json`?

I reverted to 3.2 as Magento/Framework is requiring version of `^3.2`.

**Comment by fiko** [comment]
@magento run Unit Tests

**Comment by fiko** [comment]
@magento run Static Tests, Unit Tests

**Comment by fiko** [comment]
@magento run WebAPI Tests, Semantic Version Checker, Sample Data Tests EE, Sample Data Tests CE, Sample Data Tests B2B, Magento Health Index, Integration Tests, Database Compare

**Comment by fiko** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by fiko** [comment]
@magento run Functional Tests CE

**Comment by fiko** [comment]
@magento run Functional Tests CE

**Comment by fiko** [comment]
Run Functional Tests CE 3 times, and it return errors with different results.

- 1st run, error on `AC-5262`
- 2nd run, error on `https://app.hiptest.com/projects/69435/test-plan/folders/419535/scenarios/1432482: Create an Order for a Guest Customer with a Configurable Product with Text Swatch Attributes via the Admin.`
- 3rd run, error on `MAGETWO-94111`

File: composer.json [hunk 11]
```diff
@@ -61,21 +61,23 @@
         "laminas/laminas-mvc": "^3.3",
         "laminas/laminas-oauth": "^2.4",
         "laminas/laminas-permissions-acl": "^2.10",
+        "laminas/laminas-recaptcha": "^3.6",
```
Comment (ihor-sviziev) [inline composer.json]: Why was it added? I don't think Magento has direct dependency to this package

File: composer.json [hunk 12]
```diff
@@ -61,21 +61,23 @@
...
         "laminas/laminas-servicemanager": "^3.16",
-        "laminas/laminas-soap": "^2.10",
+        "laminas/laminas-session": "^2.16",
```
Comment (ihor-sviziev) [inline composer.json]: Why was it added? I don't think Magento has direct dependency to this package

File: composer.json (hunk 9 above)
Comment (ihor-sviziev) [reply composer.json]: Feels like it's already upgraded like that in 2.4-develop, so it make sense to keep as it was.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
General feedback - looks like all changes in composer.json file we can revert and just run `composer update` to update all dependencies to versions that supports PHP 8.2.
Can't we?

**Comment by fiko** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by fiko** [comment]
@magento run all tests

**Comment by fiko** [comment]
@magento run Database Compare

**Comment by hostep** [comment]
@fiko: is the issue title & description still accurate? If not, could you update it a bit? Thanks!

**Comment by fiko** [comment]
> @fiko: is the issue title & description still accurate? If not, could you update it a bit? Thanks!

Hi @hostep 

thanks for mentioning, it's now updated as magento 2.4.6 launced few days after this PR, so previous title & description is not relevant anymore.

**Comment by fiko** [comment]
@magento run all tests

File: app/code/Magento/Reports/Test/Unit/Model/ResourceModel/Order/CollectionTest.php [hunk 13]
```diff
@@ -466,7 +466,7 @@ public function firstPartDateRangeDataProvider(): array
         return [
             ['', '', '', '0 0 0 23:59:59'],
             ['24h', '', '', '0 0 1 0:0:0'],
-            ['7d', '', '', '0 0 6 22:59:59']
+            ['7d', '', '', '0 0 6 23:59:59']
```
Comment (fredden) [inline app/code/Magento/Reports/Test/Unit/Model/ResourceModel/Order/CollectionTest.php:469]: This change is (also) being applied / fixed in https://github.com/magento/magento2/pull/37287

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @fiko,
Looks like in 2.4-develop branch we have support for PHP 8.2. Could you please verify if your changes still valid.

**Comment by fiko** [comment]
Closing this PR as the rest of the test issues are fixed on other PRs.

**Pull request closed without merging**